portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,97 +14,136 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  private HashMap<String, HashMap<String,Object>> information;  
-  private List<String> messages;
+  private HashMap<String, HashMap<String, Object>> commentInfo;
+
+  // Used in validateName() function
+  private static final String regex = "^[a-zA-Z ]+$";
+  private static final Pattern pattern = Pattern.compile(regex, Pattern.CASE_INSENSITIVE);
 
   @Override
-  public void init(){
-      messages = new ArrayList<>();
-      information = new HashMap<String, HashMap<String,Object>>();
+  public void init() {
+    commentInfo = new HashMap<String, HashMap<String, Object>>();
+    commentInfo.put("messageInfo", new HashMap<String, Object>());
+    commentInfo.put("nameInfo", new HashMap<String, Object>());
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    
-    String json = convertToJsonUsingGson(information);
+    // Query comments from DataStore as entities
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity : results.asIterable()) {
+      long id = entity.getKey().getId();
+      String name = (String) entity.getProperty("name");
+      String message = (String) entity.getProperty("message");
+      long timestamp = (long) entity.getProperty("timestamp");
+
+      Comment comment = new Comment(id, name, message, timestamp);
+      comments.add(comment);
+    }
+
+    commentInfo.get("messageInfo").put("history", comments);
+
+    String json = convertToJsonUsingGson(commentInfo);
     response.setContentType("application/json;");
     response.getWriter().println(json);
   }
 
   @Override
-  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException{
-      
-      information = checkValidity(request);
-
-      if (((Boolean) information.get("nameInfo").get("error") == true) || 
-          ((Boolean) information.get("messageInfo").get("error") == true)){
-          information.get("messageInfo").put("history", messages);
-          response.sendRedirect("/index.html#comment-page");
-      } else {
-          String message = (String) information.get("messageInfo").get("message");
-          messages.add(message);
-          information.get("messageInfo").put("history", messages);
-          response.sendRedirect("/index.html#comment-page");
-      }
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    commentInfo = checkValidity(request);
+
+    // Send invalid input to client for correction
+    if (((Boolean) commentInfo.get("nameInfo").get("error") == true)
+        || ((Boolean) commentInfo.get("messageInfo").get("error") == true)) {
```

Comment:
You don't need to check `== true` — they are already booleans.

Reply:
done